### ISSUE TITLE
Confirm status of FileSystem package

### ISSUE THREAD

**Issue by DavidWhaleMEF** [Microsoft/pxt#1682]
I have been trying the file system on pxt and pxt/beta and cannot get it to work.

Can you confirm whether the FileSystem package works yet or not?

When I run this script in the simulator it says 'pxt.sim.fileSys does not exist as a function'

When I run this on the real micro:bit and shake or press the A button, I get the heart or X then it shows a strange pattern of dots on the screen that is not in any way coded into my program.

[microbit-fs-test (5).hex.zip](https://github.com/Microsoft/pxt/files/852578/microbit-fs-test.5.hex.zip)

**Comment by pelikhan** [comment]
Until there is a firmware available to surface files to the OS, it should be considered beta. We can remove it from the "default" list.

Get Outlook for iOS<https://aka.ms/o0ukef>

________________________________

**Comment by DavidWhaleMEF** [comment]
Hmm, I'm wondering if that would be a better idea (only offer it via search) and perhaps mark it in the description as 'beta'. The problem with surfacing things that don't quite yet work (perhaps caused by secondary effects such as dependent components that are not surfaced), is that it causes a number of support tickets to come in, and quite a bit of wasted time for us and users (working out if it *should* work, and then realising it doesn't for some reason).

@jaustin might have a view on this too.

**Comment by pelikhan** [comment]
It does work on the device but it is counter intuitive that file do not appear on microbit drive.

Get Outlook for iOS<https://aka.ms/o0ukef>

________________________________

**Comment by DavidWhaleMEF** [comment]
Doesn't work for me, sorry. Crashes my micro:bit.

**Comment by pelikhan** [comment]
Could you post the shared URL?

Get Outlook for iOS<https://aka.ms/o0ukef>

________________________________

**Comment by abchatra** [comment]
+1 was moving it away from default list

**Comment by tballmsft** [comment]
yes, we should only have it there once the new firmware is out.

**Comment by pelikhan** [comment]
@finneyj looks rc8 broke our fs support. i'll investigate

**Comment by pelikhan** [comment]
The file system support broken, either with rc8 or on our side. I've removed it from the preferred package list until the supporting firmware is available.

**Comment by finneyj** [comment]
ok @pelikhan. Am building an equivalent C program to @DavidWhaleMEF's code to see if it reproduces...

**Comment by finneyj** [comment]
@pelikhan @DavidWhaleMEF 

For info, just tested this native code on microbit-dal#rc8, which (based on code inspection) is as close as I can make to the PXT generated code. I didn't test the file contents, but couldn't repro a hard crash...

Wonder if the issue is in the dimensioning of the file system? native microbit-dal will do this based on linker data, whereas PXT has a function call that takes into account PXT code and metadata... Could be in there?

```
#include "MicroBit.h"
#include "MicroBitFileSystem.h"
#include "MicroBitFile.h"

MicroBit uBit;
MicroBitFileSystem fs;

MicroBitImage heart("0,255,0,255,0,255\n255,255,255,255,255\n255,255,255,255,255\n0,255,255,255,0\n0,0,255,0,0\n");
MicroBitImage cross("255,0,0,0,255\n0,255,0,255,0\n0,0,255,0,0\n0,255,0,255,0\n255,0,0,0,255\n");
MicroBitImage triangle("0,0,0,0,0\n0,255,0,255,0\n0,0,255,0,0\n0,255,0,255,0\n255,0,0,0,255\n");

void onButtonA(MicroBitEvent)
{
    uBit.display.print(cross);
    MicroBitFile f("output.txt");
    f.remove();
    uBit.sleep(200);
    uBit.display.clear();
}

void onButtonB(MicroBitEvent)
{
    uBit.display.print(triangle);
    uBit.sleep(200);
    uBit.display.clear();
}

void onShake(MicroBitEvent)
{
    uBit.display.print(heart);
    MicroBitFile f("output.txt");
    f.append("SHAKE\n");
    f.close();
    uBit.sleep(200);
    uBit.display.clear();
}

int main()
{
    // Initialise the micro:bit runtime.
    uBit.init();

    uBit.messageBus.listen(MICROBIT_ID_GESTURE, MICROBIT_ACCELEROMETER_EVT_SHAKE, onShake); 
    uBit.messageBus.listen(MICROBIT_ID_BUTTON_A, MICROBIT_BUTTON_EVT_CLICK, onButtonA); 
    uBit.messageBus.listen(MICROBIT_ID_BUTTON_B, MICROBIT_BUTTON_EVT_CLICK, onButtonB); 

    // If main exits, there may still be other fibers running or registered event handlers etc.
    // Simply release this fiber, which will mean we enter the scheduler. Worse case, we then
    // sit in the idle task forever, in a power efficient sleep.
    release_fiber();
}
```

**Comment by pelikhan** [comment]
We do remap the file system. pxt::afterProgramPage() returns 32000).
```
    if (MicroBitFileSystem::defaultFileSystem == NULL)
    {
        /*
        char buf[128];
        sprintf(buf, "Init FS: %x %d\r\n", pxt::afterProgramPage(), pxt::programSize());
        uBit.serial.send(ManagedString(buf));
        */
        new MicroBitFileSystem(pxt::afterProgramPage());
    }
```

**Comment by finneyj** [comment]
Hmm... assuming that's hexadecimal (I'm guessing from your printf it is), that sounds plausible. Not sure how to guarantee it's correct though.

**Comment by finneyj** [comment]
I guess I could write a validation check that ensure all pages are erased before formatting a new FS...

**Comment by finneyj** [comment]
@pelikhan 

It's not well tested and certainly not for production use, but would be informative if you could target PXT at the''' filesystem-autoinit ''' branch of microbit-dal. This will scan for blank erased pages and dimension the file system accordingly.

You'll also need to change your PXT init code slightly to call the default MicrobitFileSystem constructor, rather than the one that allows the claler to specify start address and length.

if this works cleanly, I think we can safely conclude that the bug stated above is related to calculating the file system dimensions, and we can work though the best solution... Sound ok?

**Comment by pelikhan** [comment]
I will try the new branch next week.

**Comment by whaleygeek** [comment]
Any progress on this one at all?

Customer still waiting for a resolution to this at this end. Thanks!

microbit.org-ticket: 1421

**Comment by pelikhan** [comment]
Waiting for fix from Lancaster.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by finneyj** [comment]
hi guys. thanks for the prod.

The branch mentioned above contains a workaround that is tested working with PXT, but having thought about it, I don't think this is really a 100% fix. The issue is that we need to decide how to accurately dimension the FileSystem. The current (master branch) API has two ways to do this:

1) it can be defined explicitly in the constructor
2) it can automatically determine an appropriate size by using linker symbols to measure unused flash memory.

Option 2) doesn't work for PXT as user code is linked in through a different link step, so option 1) was used (but didn't work correctly)

The branch above brings in an option 3) which is to scan memory looking for blank pages to determine where the FS should be created. When used, this worked fine for PXT, highlighting that most likely, the location specified to create the file system was wrong in the current live build.

BUT i'm not sure we can simply merge that branch in. The reason being that I think it may fail if Bluetooth flashing were used, or any other technique that only performs a partial erase of memory. The reasoning being that the algorithm assumes that pages that can be used by the file system are blank erased... so any flashing approaches that might leave old blocks unerased would also result in an incorrectly dimensioned FS.

I'm wondering if the correct fix is to not merge this branch, but to do something in the JS on the PXT size to more accurately determine the place in memory where the FS should go... e.g. we could use the same/similar algorithm, if this happens at PXT compile/link time, it would be safe...

thoughs @pelikhan ?

**Comment by pelikhan** [comment]
It seems that indeed we had a size computation issue ( https://github.com/Microsoft/pxt-microbit/commit/17bab9f64f3bda409bae97780a57928237731d79 ). I've used the same strategy as in the DAL and the file system is happy again. Tests are passing on /beta, https://github.com/Microsoft/pxt-filesystem/releases/tag/v0.5.1

**Comment by finneyj** [comment]
Aawesome - that seems like a much better fix. Thanks Peli!

**Comment by pelikhan** [comment]
We traced down the issue in the DAL that was causing issues for PXT. The fix is in PXT in 0.11.9 and up and fs 0.7.1 and up.

**Comment by DavidWhaleMEF** [comment]
@pelikhan Thank you my friend, I noted the extra linking to release tags that you put into that response - much appreciated :)
